aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -0,0 +1,366 @@
+    if (!other_storage_impl->is_aliased()) {
+        // The original tensor wasn't a view - turn it into a (no-op) view.
+        other_storage_impl->set_alias(other_impl->value());
+    }
+
+    // Copy the original tensor's ViewMeta vector and push the current one.
+    if (other_impl->view_metas_.size() > 0) {
+        view_metas_ = other_impl->view_metas_;  // copy
+    }
+    view_metas_.push_back(meta);
+    storage_ = other_impl->storage_; // Ensure that storages are aliased properly.
+}
+
+// See Note [Functionalization Pass - Inplace View Ops]
+void FunctionalTensorWrapper::mutate_view_meta(at::functionalization::ViewMeta meta) {
+    auto self_impl_storage = static_cast<functionalization::FunctionalStorageImpl*>(unsafe_storage().unsafeGetStorageImpl());
+    if (!self_impl_storage->is_aliased()) {
+        // The current tensor isn't a view - turn it into one.
+        self_impl_storage->set_alias(value());
+    }
+
+    view_metas_.push_back(meta);
+    sync_(/*force_sync=*/true);
+}
+
+// Note [Functionalization: Mutation Removal]
+// Mutation removal is used to take a program like this:
+//
+// a.add_(b)
+//
+// and replace it with a slightly different program that has the same semantics:
+//
+// tmp = a.add(b)
+// a.replace_(tmp)
+//
+// Where the replace_() call is implemented directly in the functionalization pass, so it is transparent to the backend.
+// This is useful for backends that aren't able to handle certain types of mutations, like functorch.
+//
+// Why do we need to wrap every tensor in a FunctionalTensorWrapper? Consider this program:
+//
+// Before:
+// tensor.add_(batched_tensor)
+//
+// After:
+// tmp = tensor.add(batched_tensor)
+// tensor.replace_(tmp)
+//
+// In the above, tmp is a batched tensor (because adding a normal tensor to a batched tensor does broadcasting and creates a batched tensor).
+// But we can't just replace the underlying memory backing `tensor` with `tmp` - a batched tensor takes up more space!
+// Instead, every input, intermediate and output of the program is wrapped in a FunctionalTensorImpl, which wraps the underlying tensor.
+void FunctionalTensorWrapper::replace_(const Tensor& other) {
+    // TODO: going to need to change this if we want nested functionalize() transforms.
+    TORCH_INTERNAL_ASSERT(!at::functionalization::impl::isFunctionalTensor(other));
+    value_ = other;
+}
+
+
+// Note [Functionalization Pass - Inplace View Ops]
+// So, these ops are special - they're mutation AND view ops. They get special codegen.
+// An example is transpose_, e.g. `a.transpose_()`
+// Calling transpose_() should ensure that a gets an alias, and append the new ViewMeta to a's current list of ViewMetas.
+// We also need to force a sync (even if a is already up to date), because a's underlying tensor hasn't actually
+// been updated to reflect the new view yet.
+void FunctionalTensorWrapper::sync_(bool force_sync) {
```

Comment:
Is sync_ only called for inplace view ops? I can't easily tell from this comment.

Reply:
Ah no:
- `sync_()` is called by pretty much all ops except for ordinary view ops
- `sync_(/*force_sync=*/true)` is called by inplace view ops. That's because inplace/view ops don't actually need to sync the alias. They just need to re-generate themselves by replaying all of their view metas against the alias.

I should probably just make that a separate function, like `replay_from_alias()`.